pallets/sto/src/lib.rs

```diff
@@ -0,0 +1,398 @@
+                        for (id, amount) in purchases {
+                            fundraiser.tiers[id].remaining -= amount;
+                        }
+                    }
+                });
+
+                Ok(())
+            })
+        }
+
+        #[weight = 1_000]
+        pub fn freeze_fundraiser(origin, offering_asset: Ticker, fundraiser_id: u64) -> DispatchResult {
+            let sender = ensure_signed(origin)?;
+            let did = Context::current_identity_or::<Identity<T>>(&sender)?;
+            ensure!(T::Asset::primary_issuance_agent(&offering_asset) == did, Error::<T>::Unauthorized);
+
+            let now = Timestamp::<T>::get();
+            let fundraiser = <Fundraisers<T>>::get(offering_asset, fundraiser_id).ok_or(Error::<T>::FundraiserNotFound)?;
+            ensure!(fundraiser.start <= now, Error::<T>::FundraiserNotStarted);
+            if let Some(end) = fundraiser.end {
+                ensure!(end > now, Error::<T>::FundraiserExpired);
+            }
+
+            <Fundraisers<T>>::mutate(offering_asset, fundraiser_id, |fundraiser| if let Some(fundraiser) = fundraiser {
+            fundraiser.frozen = true
+            });
+            Ok(())
+        }
+
+        #[weight = 1_000]
+        pub fn unfreeze_fundraiser(origin, offering_asset: Ticker, fundraiser_id: u64) -> DispatchResult {
+            let sender = ensure_signed(origin)?;
+            let did = Context::current_identity_or::<Identity<T>>(&sender)?;
+            ensure!(T::Asset::primary_issuance_agent(&offering_asset) == did, Error::<T>::Unauthorized);
+
+            let now = Timestamp::<T>::get();
+            let fundraiser = <Fundraisers<T>>::get(offering_asset, fundraiser_id).ok_or(Error::<T>::FundraiserNotFound)?;
+            ensure!(fundraiser.start <= now, Error::<T>::FundraiserNotStarted);
+            if let Some(end) = fundraiser.end {
+                ensure!(end > now, Error::<T>::FundraiserExpired);
+            }
+
+            <Fundraisers<T>>::mutate(offering_asset, fundraiser_id, |fundraiser| if let Some(fundraiser) = fundraiser {
+            fundraiser.frozen = false
+            });
+            Ok(())
+        }
+
+        #[weight = 1_000]
+        pub fn modify_fundraiser_window(origin, offering_asset: Ticker, fundraiser_id: u64, start: T::Moment, end: Option<T::Moment>) -> DispatchResult {
+            let sender = ensure_signed(origin)?;
+            let did = Context::current_identity_or::<Identity<T>>(&sender)?;
+            ensure!(T::Asset::primary_issuance_agent(&offering_asset) == did, Error::<T>::Unauthorized);
+
+            let now = Timestamp::<T>::get();
+            let fundraiser = <Fundraisers<T>>::get(offering_asset, fundraiser_id).ok_or(Error::<T>::FundraiserNotFound)?;
+            ensure!(fundraiser.start <= now, Error::<T>::FundraiserNotStarted);
+            if let Some(end) = fundraiser.end {
+                ensure!(end > now, Error::<T>::FundraiserExpired);
```

Comment:
I don't think so we need this check here. If we are checking `end >= start` during the creation of fundraising.

Reply:
This is checking if the fundraiser expired. Even if the window is valid we still need to check for expiry.